### PR TITLE
Google Sheets Destination flushes on every state message and flush when buffer gets too large

### LINKED ISSUE #14748: Google Sheets Destination does not emit state when records are persisted.
## Problems

Consider the following section of the Google Sheets destination code:  https://github.com/airbytehq/airbyte/blob/71ab5ff21653f4e8e4341f5d5b044eb1424adc5a/airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/destination.py#L63-L73

Every time the destination receives a record, it adds it to a queue to send later.  The queue is sent only once we have 1000 records to sent to the stream (sheet).  We also immediately echo any state messages we receive. This is not what is expected for a few reasons:

1. A destination emitting a state message indicates to the platform that any preceding records for that stream have been persisted in the destination.  Since the records have only been added to a queue to be sent later, this is incorrectly signaling that those records are stored in the sheet.  This means if the sync were to crash, we won't retry those records.
2. Because the queues storing records to upload are only flushed once the queue's length is 1000, this can lead to memory issues.  If rows are particularly wide or there are many streams with sparse data, these queues won't be sent and cleared until the end of the sync (`writer.write_whats_left()`).  This ever-increasing memory usage makes an OOM crash more likely.  

Combining both issue 1 and 2 leads to an unstable destination 

## Solutions

1. State messages need to be stored and echoed only once the records that preceded those message have truly been committed to the destination.
2. We need to more regularly clear the queues to reduce the memory load of the connector

Proposal:  Every time a state message is received, we flush all the pending queues.  

```py
 for message in input_messages: 
     if message.type == Type.RECORD: 
         record = message.record 
         # process messages for available streams only 
         if record.stream in configured_stream_names: 
             writer.add_to_buffer(record.stream, record.data) 
             writer.queue_write_operation(record.stream) 
     elif message.type == Type.STATE: 
         writer.write_whats_left() # <-- Change
         yield message 
     else: 
         continue 
```

This small change will work with the legacy state messages and the new per-stream state messages.  We will be writing more, which leads to a more stable sync, but likely slower performance.

### PR DESCRIPTION
**Pull request by evantahler**
Closes https://github.com/airbytehq/airbyte/issues/14748

This PR flushes any pending RECORD writes to the Google Sheet whenever the destination sees a STATE message.  This fixes a broken spec implementation and should lead to lower memory consumption as well.

### CODE REVIEW

**Comment by evantahler** [comment]
Note for the future - the directory structure doesn't match the gradle names https://github.com/airbytehq/airbyte/pull/14751#issuecomment-1185754814

**Comment by evantahler** [comment]
The test failures are realy about permissions:

```
E   google.auth.exceptions.RefreshError: ('invalid_grant: Bad Request', ***'error': 'invalid_grant', 'error_description': 'Bad Request'***)
```

@sherifnada or @bazarnov: any tips to running these tests?

**Comment by bazarnov** [comment]
> Note for the future - the directory structure doesn't match the gradle names [#14751 (comment)](https://github.com/airbytehq/airbyte/pull/14751#issuecomment-1185754814)

What exactly should match?

**Comment by bazarnov** [comment]
> @sherifnada or @bazarnov: any tips to running these tests?

Looks like it doesn't recognise the `secret`, all tests are passed locally just now see the screenshot: 
Screenshot 2022-07-15 at 22 26 18

File: airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/destination.py [hunk 1]
```diff
@@ -69,7 +69,8 @@ def write(
                     writer.queue_write_operation(record.stream)
             elif message.type == Type.STATE:
                 # yielding a state message indicates that all preceding records have been persisted to the destination
-                writer.write_whats_left()
+                # cleaning records_buffer for target stream.
+                writer.clear_buffer(record.stream)
```
Comment (evantahler) [inline airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/destination.py:73]: @bazarnov Won't this change have the effect of erasing and not sending any records in the queues when we get a state message?  

Assuming we get messages like this:

```
RECORD
RECORD
STATE
```

We'll have 2 records in the buffer that haven't been set yet (didn't hit the 500 message threshold).  We need to `write_from_queue` before `clear_buffer`, no?

**Comment by evantahler** [comment]
Thanks for fixing the test @bazarnov, but I don't quite understand your logical change

File: airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/destination.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/destination.py]: Ah, my bad. You should be right, just because this is the `destination` not `source` and we'are working with the `source` messages. Reverted this piece.

**Comment by bazarnov** [comment]
> Thanks for fixing the test @bazarnov, but I don't quite understand your logical change

There was no change), I've simply updated the `GSM secret` to the one I've got locally, probably I forgot to update it after the successful `/publish` the initial version. (There was another link for the google sheet I believe, which was not accessible for anyone)

**Comment by evantahler** [comment]
That's great @bazarnov! If you review this PR, I'll get it published.

File: airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/writer.py [hunk 2]
```diff
@@ -34,12 +34,13 @@ def queue_write_operation(self, stream_name: str):
...
 
-        1) gets data from the records_buffer
+        1) gets data from the records_buffer, with respect to the size of the records_buffer (records count or size in Kb)
         2) writes it to the target worksheet
         3) cleans-up the records_buffer belonging to input stream
         """
-
-        if len(self.records_buffer[stream_name]) == self.flush_interval:
+        # get the size of records_buffer for target stream in Kb
+        records_buffer_size_in_kb = self.records_buffer[stream_name].__sizeof__() / 1024
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-google-sheets/destination_google_sheets/writer.py:42]: not blocking this PR over this since it's on neither of y'alls backlogs right now and is a "drive-by" change
```suggestion
        # TODO unit test flush triggers
        records_buffer_size_in_kb = self.records_buffer[stream_name].__sizeof__() / 1024
```

**Review by sherifnada — APPROVED** [review]
thanks to you both!
